utils/zerovec/src/ule/niche.rs

```diff
@@ -0,0 +1,184 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::mem::size_of;
+use core::ops::DerefMut;
+use core::{marker::Copy, ops::Deref};
+
+use super::{AsULE, ULE};
+
+/// The types implementing this trait guarantee that [`NicheBytes::INVALID_BIT_PATTERN`]
+/// can never occur as a valid byte representation of the type.
+/// The implementors guarantee that N == core::mem::sizeo_of::<Self>()
+pub trait NicheBytes<const N: usize> {
```

Comment:
Question: should this trait be `unsafe`?

I can't think of any reason it needs to be `unsafe` because if it is implemented incorrectly it just means that valid bit patterns will be incorrectly interpreted as `None` values, which isn't a safety thing. It would mean though that people depending on `NicheBytes` can't make any safety assumptions. We should at least document that.

Reply:
Done

Reply:
Why require abstractions to panic? Why panic if N is wrong, can we not compile-time check this?

Reply:
Generics in const context might cause a problem?